### DISCUSSION TITLE
How to pass username to radius server instead of local id?

### DISCUSSION THREAD

**Discussion by ttc0419** [strongswan/strongswan#1014] [Q&A] [answered]
It seems the eap-radius plugin will pass local id as username for radius request, if the local id is missing (e.g. windows does not provide a way to set local/remote id).  The plugin will send some garbage bytes as 'User-Name' to the radius server.  How can I make the plugin pass the real username to the radius server?

**example radius log:**
```
(0) Received Access-Request Id 230 from 127.0.0.1:42574 to 127.0.0.1:1812 length 158
(0)   User-Name = "\300\250\0017"
(0)   NAS-Port-Type = Virtual
(0)   Service-Type = Framed-User
(0)   NAS-Port = 1
(0)   NAS-Port-Id = "default"
(0)   NAS-IP-Address = x.x.x.x
(0)   Called-Station-Id = "x.x.x.x[4500]"
(0)   Calling-Station-Id = "y.y.y.y[41876]"
(0)   Acct-Session-Id = "1650890487-1"
(0)   EAP-Message = 0x0200000901c0a80137
(0)   NAS-Identifier = "strongSwan"
(0)   Message-Authenticator = 0x85c343645671abd4a91cf637d8a9d3f4
(0) # Executing section authorize from file /etc/freeradius/3.0/sites-enabled/strongswan
(0)   authorize {
(0) eap: Peer sent EAP Response (code 2) ID 0 length 9
(0) eap: EAP-Identity reply, returning 'ok' so we can short-circuit the rest of authorize
(0)     [eap] = ok
(0)   } # authorize = ok
(0) Found Auth-Type = eap
(0) # Executing group from file /etc/freeradius/3.0/sites-enabled/strongswan
(0)   authenticate {
(0) eap: Peer sent packet with method EAP Identity (1)
(0) eap: Calling submodule eap_mschapv2 to process data
(0) eap_mschapv2: Issuing Challenge
(0) eap: Sending EAP Request (code 1) ID 1 length 43
(0) eap: EAP session adding &reply:State = 0xc2facb59c2fbd14b
(0)     [eap] = handled
(0)   } # authenticate = handled
(0) Using Post-Auth-Type Challenge
(0) Post-Auth-Type sub-section not found.  Ignoring.
(0) Sent Access-Challenge Id 230 from 127.0.0.1:1812 to 127.0.0.1:42574 length 101
(0)   EAP-Message = 0x0101002b1a01010026101e10491990b6aad7df0fa398dc1f0e98667265657261646975732d332e302e3236
(0)   Message-Authenticator = 0x00000000000000000000000000000000
(0)   State = 0xc2facb59c2fbd14bae0a97eadd84dfdd
(0) Finished request
```

**Comment by ttc0419** [comment, marked as answer]
Okay, I found [an answer](https://serverfault.com/a/1065334) on server fault, however, changing ```eap_start``` to ```yes``` did not work for me, but luckily, adding ```eap_id = %any``` solved the issue.

**Reply by gasinvein** [reply]
Hey there, and sorry for necropositing. I've stumbled upon this discussion while searching for a similar problem.

That is, I don't think this is a proper solution, but rather a workaround. As far as I understand, setting `eap_start = no` (the default) together with `eap_id = %any`, causes strongSwan to initialize an EAP-Identity conversation with the client, and during that the client sends its EAP ID, which strongSwan uses as the `User-Name` to forward to the RADIUS server.

So, the actual issue, that strongSwan sends garbage in `User-Name`, is still there. If you set `eap_start = yes`, and the client doesn't have a manually set local id, then radius auth will fail with something like
```
(96) eap: Identity does not match User-Name, setting from EAP Identity
(96) eap: Failed in handler
(96)     [eap] = invalid
```
, which kinda makes sense, given that we have garbage in the `User-Name` attribute and probably something sensible in the EAP ID set by the client, so the values won't match each other.

What's even more strange is that the garbage-in-`User-Name` happens only with `Access-Request`, but the `Accounting-Request`s strongSwan sends have perfectly normal `User-Name` attribute (usually with the client's IP address in it).

~~I may have gaps in my understanding, but this looks like a bug on strongSwan's side.~~

**Update**: Oh well. As per documentation, the "garbage" appears to be an ASN.1 IKE identity:
>The plugin sends the IKE identity (or EAP identity if the IKE daemon is instructed to do an EAP-Identity exchange) as `User-Name` attribute to the RADIUS server.

>However, if the IKE/EAP identity is the certificate’s subject DN, matching `TLS-Client-Cert-Subject` generally won’t work because the string value it contains does not match the binary ASN.1-encoded value that’s sent in the `User-Name` attribute.
